source/blog/2016-03-11-docker-credentials-store.html.md

```diff
@@ -0,0 +1,113 @@
+To configure an *external* credentials store a user need an external helper program to interact with a specific keychain or external store. Docker requires the helper  program to be in the client's host `$PATH`. The Docker project itself is providing sane defaults for OSX and Windows. I've recently added support for a credentials store for desktop Linux ([docker/docker-credential-helpers#7](https://github.com/docker/docker-credential-helpers/pull/7)) which is using `libsecret` to work with the gnome keyring (and the KDE one).
+
+Once the helper program is installed on the client's host the user just needs to tell the Docker client to actually use it. Be aware that the user needs to perform a `docker logout` to remove the credentials already stored in the JSON file.
+
+This is what a configuration file looks like after a successful `docker login`:
+
+```bash
+$ cat $HOME/.docker/config.json
+{
+	"auths": {
+		"localhost:5001": {}
+	},
+	"credsStore": "secretservice"
+}
+```
+
+This file is telling us that we have stored credentials for the registry at `localhost:5001` and we can see no credentials are stored in the file. Instead we can see that `credsStore` is populated with the string `secretservice` (which is the default on Linux if installed). Docker reads the `credsStore` string and execute the helper `docker-credential-secretservice` to interact with the credentials store.
+
+With the help of the binary `secret-tool` we can retrieve the credentials stored (just for debugging purposes):
+
+```bash
+$ secret-tool search server localhost:5001
+[/org/freedesktop/secrets/collection/login/3774]
+label = localhost:5001
+secret = testpassword
+created = 2016-03-11 17:47:36
+modified = 2016-03-12 12:29:31
+schema = io.docker.Credentials
+attribute.username = testuser
+attribute.server = localhost:5001
+attribute.docker_cli = 1
+```
+
+When interacting with registries Docker queries the credentials store configured and uses the credentials returned (if any).
+
+Generally, it's desirable to have the system credentials store managing secrets (specifically passwords) in order to apply any additional security measures (apart from file permissions).  
+The default desktop Linux credentials store uses the system keyring to store username and password for Docker in the default Login session, in the same way that other software like Google Chrome does. This means the keyring is automatically unlocked when the user logins into the system. Passwords stored there also persist indefinitely, so that a reboot won't erase them.  
+
+The feature's pluggability helps us create and use more advanced *secrets backends*, especially for server use. We're currently exploring the [`custodia`](https://github.com/latchset/custodia) project to be used as an external credentials store for Docker, so stay tuned for more information!  
+Other implementations may store encrypted (hashed) credentials and validate the credentials provided without explicitly storing plain passwords.
```

Comment:
Yeah, fixed that, too.

Peace,
Brian

On Wed, Mar 16, 2016 at 11:49 AM, Antonio Murdaca <notifications@github.com>
wrote:

> In source/blog/2016-03-11-docker-credentials-store.html.md
> <https://github.com/projectatomic/atomic-site/pull/270#discussion_r56358773>
> :
>
> > +secret = testpassword
> > +created = 2016-03-11 17:47:36
> > +modified = 2016-03-12 12:29:31
> > +schema = io.docker.Credentials
> > +attribute.username = testuser
> > +attribute.server = localhost:5001
> > +attribute.docker_cli = 1
> > +```
> > +
> > +When interacting with registries Docker queries the credentials store configured and uses the credentials returned (if any).
> > +
> > +Generally, it's desirable to have the system credentials store managing secrets (specifically passwords) in order to apply any additional security measures (apart from file permissions).
> > +The default desktop Linux credentials store uses the system keyring to store username and password for Docker in the default Login session, in the same way that other software like Google Chrome does. This means the keyring is automatically unlocked when the user logins into the system. Passwords stored there also persist indefinitely, so that a reboot won't erase them.
> > +
> > +The feature's pluggability helps us create and use more advanced *secrets backends*, especially for server use. We're currently exploring the [`custodia`](https://github.com/latchset/custodia) project to be used as an external credentials store for Docker, so stay tuned for more information!
> > +Other implementations may store encrypted (hashed) credentials and validate the credentials provided without explicitly storing plain passwords.
>
> Thank you, btw I still see the bash tags, see #270 (comment)
> <https://github.com/projectatomic/atomic-site/pull/270#discussion_r56343027>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/projectatomic/atomic-site/pull/270/files/d103b1c4f734e2ed03f1aeab039e4812fd31a14d#r56358773>
>



-- 
Brian Proffitt
Principal Community Analyst
Open Source and Standards
@TheTechScribe
574.383.9BKP